clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/VeniceAdaptiveBlobTransferTrafficThrottler.java

```diff
@@ -0,0 +1,171 @@
+package com.linkedin.davinci.blobtransfer;
+
+import com.linkedin.venice.throttle.VeniceAdaptiveThrottler;
+import io.netty.handler.traffic.GlobalChannelTrafficShapingHandler;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BooleanSupplier;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * Adaptive throttling wrapper for {@link io.netty.handler.traffic.GlobalChannelTrafficShapingHandler}
+ * that dynamically tunes global read or write limits for blob transfer based on registered boolean signals.
+ * It controls either global read or write throughput for the blob transfer behavior based on results of the registered
+ * limiter and booster signals.
+ *
+ * The heuristic behavior rules are defined as:
+ * (1) If any limiter signal found, it will decrease by 20%. (min 20% of the base rate).
+ * (2) If no limiter signal found, but there is booster signal found, it will increase by 20%. (max 200% of the base rate).
+ * (3) If none of the above rule applies, it will increase idle count, if idle count is greater than the pre-defined
+ * threshold, it will try to increase by 20% (still max 200% of the base rate).
+ *
+ * Base rate is initially set to {@link com.linkedin.venice.ConfigKeys#BLOB_TRANSFER_SERVICE_WRITE_LIMIT_BYTES_PER_SEC}
+ * or {@link com.linkedin.venice.ConfigKeys#BLOB_TRANSFER_CLIENT_READ_LIMIT_BYTES_PER_SEC} and will be adjusted based on
+ * input signals dynamically.
+ */
+public class VeniceAdaptiveBlobTransferTrafficThrottler implements VeniceAdaptiveThrottler {
+  private static final Logger LOGGER = LogManager.getLogger(VeniceAdaptiveBlobTransferTrafficThrottler.class);
+  private static final String THROTTLER_NAME_SUFFIX = "TrafficBlobTransferThrottler";
+  private final String throttlerName;
+  private final List<BooleanSupplier> limiterSuppliers = new ArrayList<>();
+  private final List<BooleanSupplier> boosterSuppliers = new ArrayList<>();
+  private final AtomicReference<GlobalChannelTrafficShapingHandler> globalChannelTrafficShapingHandler =
+      new AtomicReference<>();
+  private final long baseRate;
+  private final int signalIdleThreshold;
+  private int currentFactor = 100;
+  private static final int MIN_FACTOR = 20;
+  private static final int MAX_FACTOR = 200;
+  private final int rateUpdatePercentage;
+  private final boolean isWriteThrottler;
+  private int signalIdleCount;
+
+  public VeniceAdaptiveBlobTransferTrafficThrottler(
+      int singleIdleThreshold,
+      long baseRate,
+      int rateUpdatePercentage,
+      boolean isWriteThrottler) {
+    this.throttlerName = (isWriteThrottler ? "Write" : "Read") + THROTTLER_NAME_SUFFIX;
```

Comment:
nit: can you a delimiter like "-" or "_"

Reply:
Actually this is intended, I would like it to be looking like:
**ReadTrafficBlobTransferThrottler**
and
**WriteTrafficBlobTransferThrottler**